### PR TITLE
[SYCL][E2E]  Update get_coordinate tests to verify both row and columns

### PR DESCRIPTION
**Pull request by YixingZhang007**
Improving Test Coverage and Cleanliness:
- Implement validation for the `get_coordinate` function to check both row and column indices for all matrix types in `get_coordinate_ops.cpp`.
- Remove the original tests: `get_coord_int8_matA.cpp`, `get_coord_int8_matB.cpp`, and `get_coord_float_matC.cpp`. These tests have been consolidated and are now included in `get_coordinate_ops.cpp`.

### CODE REVIEW

**Comment by YixingZhang007** [comment]
This PR is still in progress. There might be some code clean up needed to be done.

**Comment by uditagarwal97** [comment]
@YixingZhang007 I've updated the PR description to remove the JIRA link. The policy does not allow using internal JIRA links in open-source, only JIRA ticket numbers are allowed.

File: sycl/test-e2e/Matrix/get_coord_int8_matB_impl.hpp [hunk 1]
```diff
@@ -165,6 +182,76 @@ void matrix_sum_cols(big_matrix<T, Rows, Cols> &B,
                                        sum_cols_v.get_host_access());
 }
 
+template <typename T, typename TResult, size_t Rows, size_t Cols,
+          size_t TileRows, size_t TileCols, size_t VNNI>
+void matrix_sum_rows(big_matrix<T, Rows, Cols> &B,
```
Comment (dkhaldi) [inline sycl/test-e2e/Matrix/get_coord_int8_matB_impl.hpp:187]: Minor:
Is there a way to include matrix_sum_rows as part of matrix_sum_cols with may be an extra argument to specify whether we want to sum the rows or cols?

Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coord_int8_matB_impl.hpp]: Thank you for the suggestion! I have combined `matrix_sum_rows` and `matrix_sum_cols` into a single function called `matrix_sum`.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+               aref.fetch_add(sum_local_cols[i]);
+             }
+           }
+         }); // parallel for
+   }).wait();
+}
+
+template <typename T, typename TResult, size_t SROWS, size_t SCOLS, use Use, 
+        layout Layout, size_t VF>
+void test_gcop() {
```
Comment (dkhaldi) [inline sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:164]: what does "gcop" stand for?
It is better to spell it out.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                   aref(v_cols[i]);
+               aref.fetch_add(sum_local_cols[i]);
+             }
+           }
+         }); // parallel for
+   }).wait();
+}
+
+template<typename T, typename TResult, size_t NUM_ROWS, size_t NUM_COLS, size_t SROWS, size_t SCOLS, use Use, layout Layout, size_t VF>
+void matrix_sum_ab(big_matrix<T, NUM_ROWS / VF, NUM_COLS * VF> &M, TResult *sum_rows, TResult *sum_cols) {
```
Comment (dkhaldi) [inline sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: why matrix_sum_c and matrix_sum_ab are not in one function?

Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: Matrix C and Matrices A/B are tested in two different `matrix_sum` functions due to the following reasons:

1. When calling `joint_matrix_load` for Matrix C, a layout must be specified. For Matrices A and B, including a layout results in an error.

Example for Matrix C:
```
joint_matrix_load(
               sg, sub,
               acc.template get_multi_ptr<access::decorated::no>() +
                   (sg_startx * SROWS * NUM_COLS) + sg_starty / sg_size * SCOLS,
               NUM_COLS, Layout);
```

Example for Matrices A and B:
```
joint_matrix_load(
               sg, sub,
               acc.template get_multi_ptr<access::decorated::no>() +
                   (sg_startx * (SROWS / VF) * NUM_COLS * VF) +
                   sg_starty / sg_size * SCOLS * VF,
               NUM_COLS * VF);
```

2. For Matrices A and B, the layout must be passed as a template argument when calling the `joint_matrix` function. For Matrix C, including layout will result in errors.  

Example for Matrix C:
```
joint_matrix<sub_group, T, Use, SROWS, SCOLS> sub;
```

Example for Matrices A and B:
```
joint_matrix<sub_group, T, Use, SROWS, SCOLS, Layout> sub;
``` 

For a similar example demonstrating the use of different functions for Matrices A/B and Matrix C, please refer to this link: https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp

File: sycl/test-e2e/Matrix/get_coordinate_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
+
+
+
+//==----------- get_coord_float_matC_impl.hpp  - DPC++ joint_matrix---------==//
```
[the remark this answers is not in the document]
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_impl.hpp]: Sorry for the mistake! I have fixed it. Thanks!

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp (hunk 2 above)
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: Thanks for the suggestion! I have changed it.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+  std::vector<combination> combinations =
+      q.get_device()
+          .get_info<sycl::ext::oneapi::experimental::info::device::
+                        matrix_combinations>();
+
+  for (unsigned int i = 0; i < combinations.size(); i++) {
+    if (combinations[i].nsize == 0) { // Intel AMX
+      test_gcop<int8_t, int, /*TM*/ 16, /*TK*/ 64, use::a, layout::row_major, 1>();
+      test_gcop<int8_t, int32_t, /*TK*/ 64, /*TN*/ 16, use::b, layout::ext_intel_packed, 4>();
+      test_gcop<float, float, /*TM*/ 16, /*TN*/ 16, use::accumulator, layout::row_major, 1>();
```
[the remark this answers is not in the document]
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: Thanks for the suggestions! I have added the following two test cases for intel_gpu_pvc.
```
test_get_coord_op<bfloat16, float, /*TM*/ 8, /*TK*/ 16, use::a,
                        layout::row_major, 1>();
test_get_coord_op<bfloat16, float, /*TK*/ 16, /*TN*/ 16, use::b,
                        layout::ext_intel_packed, 2>();
```

**Comment by YixingZhang007** [comment]
The CI test, `E2E tests with dev igc on Intel Arc A-Series Graphics`, is currently failing. When I ran the E2E tests on a DG2 machine, all tests passed without any issues. I am still investigating the cause of this failure.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp (hunk 3 above)
Comment (dkhaldi) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: `matrix_sum` has more common code.
Can you pass both layout of A/b as template parameter and layout of C as function argument? Then inside the function, call the appropriate function based on value of `use`.
The other suggestion  is to move the common code after joint_matrix_apply to a common function. 
Can either of these work?

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkhaldi) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: Please add also:
 test_get_coord_op<bfloat16, float, /*TK*/ 16, /*TN*/ 16, use::b, layout::row_major, 1>();

[the remark this answers is not in the document]
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: I’ve added the following test for Matrix B Row_Major on intel_gpu_dg2:
`test_get_coord_op<bfloat16, float, /*TK*/ 16, /*TN*/ 8, use::b, layout::row_major, 1>();`

I wasn’t able to add the test `test_get_coord_op<bfloat16, float, /*TK*/ 16, /*TN*/ 16, use::b, layout::row_major, 1>();` for intel_gpu_pvc, as we already have the test `test_get_coord_op<bfloat16, float, /*TK*/ 16, /*TN*/ 16, use::b, layout::ext_intel_packed, 2>();`. Including both tests would result in a conflict due to a duplicate definition for `big_matrix<bfloat16, 16, 16>`, which generates the following error
```
../../get_coordinate_ops_impl.hpp:29:10: error: definition with same mangled name '_ZTS10big_matrixIN4sycl3_V13ext6oneapi8bfloat16ELm32ELm32EE' as another definition
   29 |          [=](nd_item<2> spmd_item)
      |          ^
../../get_coordinate_ops_impl.hpp:29:10: note: previous definition is here
```

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp (hunk 3 above)
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: Thanks for the suggestion :) I’ve updated the `matrix_sum` function to include a simplified implementation. Now, the following code is used for both Matrix A/B and Matrix C:
```
joint_matrix<sub_group, T, Use, SROWS, SCOLS, Layout> sub;
           joint_matrix_load(
               sg, sub,
               acc.template get_multi_ptr<access::decorated::no>() +
                   (sg_startx * (SROWS / VF) * NUM_COLS * VF) +
                   sg_starty / sg_size * SCOLS * VF,
               NUM_COLS * VF);
```

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkhaldi) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: >>> Including both tests would result in a conflict due to a duplicate definition for big_matrix<bfloat16, 16, 16>, which generates the following error

The solution to this is to add a differentiating argument (layout in this case) to
`big_matrix<T, Rows / VF, Cols * VF> MM((T *)&Mvnni);`

[the remark this answers is not in the document]
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp]: Thanks! I have resolved the issue with introducing a new class, `template <size_t Rows, size_t Cols, layout Layout, use Use> class matrix_process;`, to differentiate between matrix tests.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+      test_get_coord_op<int8_t, int, /*TM*/ 8, /*TK*/ 32, use::a,
+                        layout::row_major, 1>();
+      test_get_coord_op<bfloat16, float, /*TK*/ 16, /*TN*/ 16, use::b,
+                        layout::row_major, 1>();
+      test_get_coord_op<bfloat16, float, /*TK*/ 16, /*TN*/ 16, use::b,
+                        layout::ext_intel_packed, 2>();
+      test_get_coord_op<int8_t, int32_t, /*TK*/ 32, /*TN*/ 16, use::b,
+                        layout::ext_intel_packed, 4>();
+      test_get_coord_op<float, float, /*TM*/ 8, /*TN*/ 16, use::accumulator,
+                        layout::row_major, 1>();
```
Comment (dkhaldi) [inline sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:184]: Awesome, please add the extra cases to DG2 and AMX as well: 
- a row major bfloat16
- b row major bfloat16
- b packed bfloat16
- b row major int8
- c int32 row major 

Also, I see that - b row major int8 is missing in PVC. Please add that as well

Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:192]: Thanks for the suggestion! I have included each of the following cases for DG2, AMX  and PVC. 
- a row_major  bloat16 
- a row_major  int8
- b row_major bloat16
- b row_major int8
- b packed bloat16
- b packed int8
- a float
- a int32

Please feel free to let me know if there is any other cases need to be added :)

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+  matrix_vnni<T>(Rows, Cols, *M, *Mvnni, VF);
+  big_matrix<T, Rows / VF, Cols * VF> MM((T *)&Mvnni);
+
+  matrix_sum<T, TResult, Rows, Cols, SROWS, SCOLS, Use, Layout, VF>(
+      MM, sum_rows, sum_cols);
+
+  // This condition check can be removed once the IGC PR resolving the Matrix B
+  // row coordinate bug is pull downed to the driver.
+  if (Use != use::b) {
```
Comment (dkhaldi) [inline sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:141]: this is not needed anymore

**Review by dkhaldi — APPROVED** [review]
LGTM

**Comment by dkhaldi** [comment]
Ping @intel/llvm-reviewers-runtime

**Comment by YixingZhang007** [comment]
@intel/llvm-gatekeepers This PR has been approved and all CI tests passed. Could you please help merge this PR? Thanks!

**Review by againull — APPROVED** [review]

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+             sum_local_cols[i] =
+                 reduce_over_group(sg, sum_local_cols[i], sycl::plus<>());
+             // only Groups leader perform the global reduction
+             if (global_idy % sg_size == 0) {
+               sycl::atomic_ref<TResult, sycl::memory_order::relaxed,
+                                sycl::memory_scope::device>
+                   aref(v_cols[i]);
+               aref.fetch_add(sum_local_cols[i]);
+             }
+           }
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:108]: @YixingZhang007 this code is duplicated. Would be great to outline it to function and reuse.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+  TResult sum_rows[Rows] = {0};
+  TResult sum_rows_ref[Rows] = {0};
+  TResult sum_cols[Cols] = {0};
+  TResult sum_cols_ref[Cols] = {0};
+
+  for (int i = 0; i < Rows; i++) {
+    for (int j = 0; j < Cols; j++) {
+      M[i][j] = i + j;
+    }
+  }
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:131]: @YixingZhang007 , we have `matrix_fill` in `common.hpp` to not duplicate such loops in each test.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp (hunk 8 above)
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:108]: Thanks for the suggestion! I have created a PR for fixing this. The link to the PR is https://github.com/intel/llvm/pull/16429

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp (hunk 9 above)
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:131]: Thanks! I have changed this.

File: sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+     sycl::accessor acc{buf, cgh, sycl::read_write};
+     sycl::accessor v_rows{sum_rows_v, cgh, sycl::read_write};
+     sycl::accessor v_cols{sum_cols_v, cgh, sycl::read_write};
+
+     cgh.parallel_for<matrix_process<T, NUM_ROWS, NUM_COLS, Layout, Use>>(
+         nd_range<2>({NUM_ROWS / SROWS, NUM_COLS / SCOLS * sg_size},
+                     {1, 1 * sg_size}),
+         [=](nd_item<2> spmd_item)
+#ifdef SG_SZ
+             [[intel::reqd_sub_group_size(SG_SZ)]]
```
[the remark this answers is not in the document]
Comment (YixingZhang007) [reply sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp:35]: Sorry for the mistake! I have fix this in the other PR.

**Pull request merged**